llpc/util/llpcCacheAccessor.h

```diff
@@ -0,0 +1,132 @@
+ * @file  llpcUtil.h
+ * @brief LLPC header file: contains the definition of LLPC internal types and utility functions
+ * (independent of LLVM use).
+ ***********************************************************************************************************************
+ */
+#pragma once
+
+#include "llpc.h"
+#include "llpcShaderCache.h"
+#include "vkgcMetroHash.h"
+
+namespace Llpc {
+
+class Compiler;
+class Context;
+
+class CacheAccessor {
+public:
+  // ===================================================================================================================
+  // Checks the caches in the build info and compiler objects for an entry with the given hash.
+  //
+  // @param buildInfo : The build information that will give the caches from the application.
+  // @param hash : The hash for the entry to access.
+  // @param compiler : The compiler object with the internal caches.
+  template <class BuildInfo> CacheAccessor(BuildInfo *buildInfo, MetroHash::Hash &cacheHash, Compiler *compiler) {
+    initializeUsingBuildInfo(buildInfo, cacheHash, compiler);
+  }
+
+  CacheAccessor(Context *context, MetroHash::Hash &cacheHash, Compiler *compiler);
+
+  // ===================================================================================================================
+  // Finalizes the cache access by releasing any handles that need to be released.
+  ~CacheAccessor() { setElfInCache({0, nullptr}); }
+
+  // ===================================================================================================================
+  // Returns true of the entry was in at least on of the caches or has been added to the cache.
+  bool isInCache() const {
+    return m_cacheResult == Result::Success || m_shaderCacheEntryState == ShaderEntryState::Ready;
+  }
+
+  // ===================================================================================================================
+  // Returns the ELF that was found in the cache.
+  BinaryData getElfFromCache() const { return m_elf; }
+
+  void setElfInCache(BinaryData elf);
+
+  // ===================================================================================================================
+  // Returns true if there was a cache hit in an internal cache.
+  bool hitInternalCache() const {
+    if (!isInCache())
+      return false;
+    if (m_cacheResult == Result::Success) {
+      return true;
+    }
+    return getUserShaderCache() == m_shaderCache;
+  }
+
+private:
+  Vkgc::ICache *getUserCache() const { return m_userCache; }
+  IShaderCache *getUserShaderCache() const { return m_userShaderCache; }
+
+  // ===================================================================================================================
+  // Initializes all of member variable using.
+  template <class BuildInfo>
+  void initializeUsingBuildInfo(const BuildInfo *buildInfo, MetroHash::Hash &hash, Compiler *compiler) {
+    Vkgc::ICache *userCache = buildInfo->cache;
```

Comment:
Could we add assertions for pointer arguments that should not be null? Or turn them into references.